### DISCUSSION TITLE
Reverse Proxy HTTPS and Basic Authentication to Theia for use on iPadPro

### DISCUSSION THREAD

**Discussion by maximsachs** [eclipse-theia/theia#9397] [Q&A]
### Description
Hello,
Im trying to set up all my tools to be completely browser based for a mobile and secure office experience, while using a light device to interact with everything. In my case specifically I am using a iPadPro with iPadOs beta to make use of the extended browser capabilities.

So far I have got most things working, the last hang up unfortunately is the Theia-IDE. Theia is working great, also on the iPad, however when trying to make things more secure things stop working. 

With https and http-basic authentication, Chrome on my desktop is able to access theia fine, but on the iPadPro, the spinning loading icon appears and then just spins forever, instead of loading the workspace etc.

Heres what I found out so far:

### Reproduction Steps
I used this guide https://www.theia-ide.org/doc/composing_applications to build Theia on my computer at home which is running **Ubuntu 16.04 LTS Desktop**. Running Theia locally works like a charm. When I am on the road, I use openVpn to connect home, which then allows me to access the different browser based tools. However that is not good enough, so using Letsencrypt I have set up certificates to switch all tools to use https.

Since I couldnt find any internal https support for Theia I followed the advice from other issues to set up a reverse proxy using **NGINX**. 

This is however not sufficient still in my opinion, as an attacker as long as they are in the local network could easily access my entire computer through the terminal that Theia provides. Therefore I wanted to add http basic authentication to the reverse proxy. This would mean an attacker would have to bypass the vpn, or have dirrect physical access to the local network, then would have to intercept the https traffic to be able to get the username and password, which is hard because my local network is configured to use its own DNS server which is secured itself. So in my opinion that is sufficiently secure against a casual attacker, or accidental access by friends on my local network.

This is my NGINX configuration. (Theia is running on localhost with port 8126):
`/etc/nginx/nginx.conf`:
```
user www-data;
worker_processes auto;
pid /run/nginx.pid;

events {
	worker_connections 768;
	# multi_accept on;
}

http {
	map $http_upgrade $connection_upgrade {
		default upgrade;
		upgrade close;
	}

	##
	# Basic Settings
	##

	sendfile on;
	tcp_nopush on;
	tcp_nodelay on;
	keepalive_timeout 65;
	types_hash_max_size 2048;
	# server_tokens off;

	# server_names_hash_bucket_size 64;
	# server_name_in_redirect off;

	include /etc/nginx/mime.types;
	default_type application/octet-stream;

	log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

	##
	# SSL Settings
	##

	ssl_protocols TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
	ssl_prefer_server_ciphers on;

	##
	# Logging Settings
	##

	access_log /var/log/nginx/access.log;
	error_log /var/log/nginx/error.log;

	##
	# Gzip Settings
	##

	gzip on;
	gzip_disable "msie6";

	gzip_vary on;
	gzip_proxied any;
	gzip_comp_level 6;
	gzip_buffers 16 8k;
	gzip_http_version 1.1;
	gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

	##
	# Virtual Host Configs
	##

	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
}
```

And `/etc/nginx/sites-enabled/default`:
```
# Default server configuration
#
server {
	listen                          80;
	server_name                     https://theia.cryptobro.eu:41349;
	return 301                      $server_name$request_uri;
}

server {
	# SSL configuration
	#
	listen 41349 ssl default_server;
	listen [::]:41349 ssl default_server;
	ssl	on;
	ssl_certificate /etc/letsencrypt/live/theia.cryptobro.eu/fullchain.pem;
	ssl_certificate_key /etc/letsencrypt/live/theia.cryptobro.eu/privkey.pem;

	server_name theia.cryptobro.eu;

	location / {
		proxy_pass http://localhost:8126;
		proxy_set_header Host $host;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_http_version 1.1;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection $connection_upgrade;
		add_header Front-End-Https on;
		auth_basic                            "Username and Password Required";
                auth_basic_user_file                  /etc/nginx/.htpasswd;

		#client_max_body_size 0;
		#add_header Strict-Transport-Security "max-age=63072000; includeSubDomains; preload" always;
		#add_header Referrer-Policy "same-origin" always;
        }
}
```

The configuration above works well on my normal computer with chrome (also ubuntu 16.04). It is supposed to use https for any request (an http request should automatically be changing to https), and additionally send the http basic authentication with each request. Then they are supposed to be forwarded internally to the theia server running on localhost. 

**The Problem**

While chrome on ubuntu has no issues, the iPadPro doesnt seem to like this set up. When connecting it promts for the login credentials for basic authentication, which are entered and if correct proceeds to load the website. It is getting the server itself, however then it is stuck on the theia loading icon, which is spinning forever. After some trial and error with different settings, I think the problem is something to do with the connection upgrade for the websocket connection. If I understand correctly, the websocket is what gets the actual files and such from the current workspace open in theia. It would make sense that if everything but the websocket is accessible, that the website partially loads and is then stuck on the loading icon, because it gets some error when trying to get data from the websocket. From some googling I found that Apple apparently made some harsh choices to improve security with http-basic authentication, which is not allowed in safari unless its going over a https connection. I have a feeling that the problem lies somewhere there, that the websocket is not correctly upgraded to a secure connection, which results in safari on the iPad not including the correct http-basic authentication, which then results in unauthenticated error on the reverse proxy. I am not sure if this is a problem of my configuration, or if the theia frontend for some reason is sending requests that are not upgraded to https or maybe another cause.

For testing I also tried to disable the http-basic authentication in the reverse proxy, however still using ssl. That works just fine to load and use theia on the ipad, but with http-basic auth it doesnt work anymore.

 Any help would be greatly appreciated. I tried to make this as detailed as possible. Thank you all very much.

**Is my reverse proxy configuration correct?**

**Diagnostics:** 
Nginx access log:
```
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /e817279537a0417d042f62fbb1b99eea.gif HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET / HTTP/1.1" 200 175 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /vs/editor/editor.main.css HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /vs/loader.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /vs/editor/editor.main.nls.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /vs/language/html/monaco.contribution.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /vs/language/css/monaco.contribution.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /vs/editor/editor.main.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /f8f821a66b20e5801e35df5ec4565b30.wasm HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /bundle.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:00 +0200] "GET /bundle.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /e817279537a0417d042f62fbb1b99eea.gif HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /vs/loader.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /vs/editor/editor.main.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /vs/editor/editor.main.css HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /vs/editor/editor.main.nls.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /e817279537a0417d042f62fbb1b99eea.gif HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /vs/language/css/monaco.contribution.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /vs/language/html/monaco.contribution.js HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /f8f821a66b20e5801e35df5ec4565b30.wasm HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:01 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - maxim [29/Jul/2019:14:07:01 +0200] "GET /e817279537a0417d042f62fbb1b99eea.gif HTTP/1.1" 304 0 "https://theia.cryptobro.eu:41349/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:03 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:06 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:09 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:13 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:19 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:27 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:36 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:46 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:07:56 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:08:06 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:08:16 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:08:26 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:08:36 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:08:46 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
192.168.1.62 - - [29/Jul/2019:14:08:57 +0200] "GET /services HTTP/1.1" 401 204 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15"
```

**Comment by akosyakov** [comment]
@marcdumais-work @geropl @32leaves @AlexTugarev maybe you have some insights? 

I don't have enough exp with nginx to answer.

**Comment by uniibu** [comment]
@maximsachs  Firstly, you cannot do `server_name https://theia.cryptobro.eu:41349;` https://nginx.org/en/docs/http/server_names.html. I'm not sure what is the `41349` port for? Non-standard https port?

Anyway, for the first part(http) it should simply be like this:

```
server {
        listen 80;
        listen [::]:80;
        server_name theia.cryptobro.eu;
        return 301 https://$host$request_uri;
}
```

The above will redirect(301) all http connection to your site to https.
For the the 2nd part(https) it would be like this:

```
server {
        listen 443 ssl http2;
        listen [::]:443 ssl http2;
        server_name theia.cryptobro.eu;
        ssl_certificate /etc/letsencrypt/live/theia.cryptobro.eu/fullchain.pem;
	ssl_certificate_key /etc/letsencrypt/live/theia.cryptobro.eu/privkey.pem;
        autoindex off;
        charset utf-8;

       # this limits access to the whole site with basic authentication
        auth_basic  "Username and Password Required";
        auth_basic_user_file  /etc/nginx/.htpasswd;

       location / {
             proxy_set_header Host $http_host;
             proxy_set_header X-Real-IP $remote_addr;
             proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
             proxy_set_header X-Forwarded-Proto $scheme;
             client_max_body_size 100M;
             client_body_buffer_size 1m;
             proxy_http_version 1.1;
	     proxy_redirect off;

             # start of optimizations, could be adjusted/ignored
	     proxy_intercept_errors on;
	     proxy_buffering on;
	     proxy_buffer_size 128k;
	     proxy_buffers 256 16k;
	     proxy_busy_buffers_size 256k;
	     proxy_temp_file_write_size 256k;
	     proxy_max_temp_file_size 0;
	     proxy_read_timeout 300;
             # end of optimizations

             proxy_set_header Upgrade $http_upgrade;
             proxy_set_header Connection "upgrade";
             proxy_pass http://localhost:8126;
       }
}
```
And lastly remove below code from your nginx.conf
```
map $http_upgrade $connection_upgrade {
		default upgrade;
		upgrade close;
	}
```

Here's a simple [gist](https://gist.github.com/uniibu/e6dfab6ac08ccdeddad2bbe9f22feae1) which i basically use(except the basic_ath, I use cloudflare access for authentication)

**Comment by khannurien** [comment]
I too get the spinning wheel using SSL/Basic Auth with nginx as reverse proxy and Safari 13.0.2.

`[Error] WebSocket connection to 'wss://mydomain.com/services' failed: Unexpected response code: 401`

**Comment by trayla** [comment]
I can acknowledge this problem, too. Both iOS and Safari Browsers cannot show Theia with activated basic authentication in the reverse proxy. I am using a Kubernetes Ingress based on nginx.

**Comment by shenjc2013** [comment]
Me too,  theia with nginx auth is Ok in MACOS and WIN,But it is some wrong with ipados and IOS

**Comment by shenjc2013** [comment]
In the end, I dealt with the problem,But wss location is not with basic auth；
here is example:

server {
	server_name code.kuber.top;
	listen 80 ;
	access_log /var/log/nginx/access.log vhost;
	include /etc/nginx/vhost.d/default;
	location / {
                auth_basic_user_file auth.db;
                auth_basic "Restricted";
		proxy_pass http://code.kuber.top;
	}

	location /services {
		proxy_pass http://code.kuber.top;
	}
}
server {
	server_name code.kuber.top;
	listen 443 ssl http2 ;
	access_log /var/log/nginx/access.log vhost;
        ssl_certificate      ./certs/kuber.top.pem;
        ssl_certificate_key  ./certs/kuber.top.key;
        auth_basic_user_file auth.db;
        auth_basic "Restricted";
	location / {
		proxy_pass http://code.kuber.top;
	}

	location /services {
		proxy_pass http://code.kuber.top;
	}
}

**Comment by maximsachs** [comment]
I am also using the above mentioned solution of https for everything except the websocket. Additionally now after an update of the theia plugins, the pdf files for example would no longer be shown inside the browser. 

I identified the issue to be due to the mini-browser plugin. By default it uses a host pattern of "{{uuid}}.mini-browser.{{hostname}}". This of course does not work with the reverse proxy, first because the uuid is a different subdomain everytime, and secondly even without it a second ssl certificate would be required for the mini-browser. domain. 

what worked for me is setting the following environment variable before running theia and then the existing reverse proxy config continues to work just fine:

`export THEIA_MINI_BROWSER_HOST_PATTERN={{hostname}}`

**Comment by paul-marechal** [comment]
Note that while it works to set `THEIA_MINI_BROWSER_HOST_PATTERN={{hostname}}`, it is recommended to get your DNS resolvers to properly resolve `*.mini-browser.*` domains to the same IP where the Theia backend is running. This is called virtual hosting and it is done to isolate each mini-browser views on their own origins for security purposes. I cannot help you setting up such a configuration beside this high level information, as DNS configurations vary based on the kind of DNS service you happen to be using.

Theia's webviews use the same mechanism.
